bson/src/test/unit/org/bson/json/JsonReaderTest.java

```diff
@@ -328,673 +458,1046 @@ public void testNumberInt() {
+                return null;
+            }
+        });
     }
 
     @Test
     public void testDBPointer() {
         String json = "DBPointer(\"b\",\"5209296cd6c4e38cf96fffdc\")";
-        bsonReader = new JsonReader(json);
-        assertEquals(BsonType.DB_POINTER, bsonReader.readBsonType());
-        BsonDbPointer dbPointer = bsonReader.readDBPointer();
-        assertEquals("b", dbPointer.getNamespace());
-        assertEquals(new ObjectId("5209296cd6c4e38cf96fffdc"), dbPointer.getId());
+        testStringAndStream(json, new Function<AbstractBsonReader, Void>() {
+            @Override
+            public Void apply(final AbstractBsonReader bsonReader) {
+                assertEquals(BsonType.DB_POINTER, bsonReader.readBsonType());
+                BsonDbPointer dbPointer = bsonReader.readDBPointer();
+                assertEquals("b", dbPointer.getNamespace());
+                assertEquals(new ObjectId("5209296cd6c4e38cf96fffdc"), dbPointer.getId());
+                return null;
+            }
+        });
     }
 
     @Test
     public void testDBPointerWithNew() {
         String json = "new DBPointer(\"b\",\"5209296cd6c4e38cf96fffdc\")";
-        bsonReader = new JsonReader(json);
-        assertEquals(BsonType.DB_POINTER, bsonReader.readBsonType());
-        BsonDbPointer dbPointer = bsonReader.readDBPointer();
-        assertEquals("b", dbPointer.getNamespace());
-        assertEquals(new ObjectId("5209296cd6c4e38cf96fffdc"), dbPointer.getId());
+        testStringAndStream(json, new Function<AbstractBsonReader, Void>() {
+            @Override
+            public Void apply(final AbstractBsonReader bsonReader) {
+                assertEquals(BsonType.DB_POINTER, bsonReader.readBsonType());
+                BsonDbPointer dbPointer = bsonReader.readDBPointer();
+                assertEquals("b", dbPointer.getNamespace());
+                assertEquals(new ObjectId("5209296cd6c4e38cf96fffdc"), dbPointer.getId());
+                return null;
+            }
+        });
+    }
+
+    @Test
+    public void testMultipleMarks() {
+        String json = "{a : { b : 1 }}";
+        testStringAndStream(json, new Function<AbstractBsonReader, Void>() {
+            @Override
+            public Void apply(final AbstractBsonReader bsonReader) {
+                bsonReader.readStartDocument();
+                BsonReaderMark markOne = bsonReader.getMark();
+                bsonReader.readName("a");
+                bsonReader.readStartDocument();
+                BsonReaderMark markTwo = bsonReader.getMark();
+                bsonReader.readName("b");
+                bsonReader.readInt32();
+                bsonReader.readEndDocument();
+                markTwo.reset();
+                bsonReader.readName("b");
+                markOne.reset();
+                bsonReader.readName("a");
+                return null;
+            }
+        });
+    }
+
+    @Test
+    public void testTwoDocuments() {
```

Comment:
I like that we can do this now.  It means we can easily parse the output of mongoexport

Reply:
ACK